core/src/preferences/PreferencesBackend.cpp

```diff
@@ -194,17 +194,47 @@ namespace ledger {
             unsetEncryption();
 
             auto emptySalt = std::string("");
-            auto pref = getPreferences("__core");
-            auto salt = pref->getString(ENCRYPTION_SALT_KEY, emptySalt);
+            auto saltKey = std::vector<uint8_t>(ENCRYPTION_SALT_KEY.cbegin(), ENCRYPTION_SALT_KEY.cend());
+            auto salt = get(saltKey).value_or(emptySalt);
 
             if (salt == emptySalt) {
-                // we don’t have a proper salt; create one and persist it for future use
+                leveldb::WriteBatch batch;
+
+                // we don’t have a proper salt; create one
                 salt = createNewSalt(rng);
-                pref->editor()->putString(ENCRYPTION_SALT_KEY, salt)->commit();
-            }
 
-            // create the AES cipher
-            _cipher = AESCipher(rng, password, salt, PBKDF2_ITERS);
+                // create the AES cipher
+                _cipher = AESCipher(rng, password, salt, PBKDF2_ITERS);
+
+                // add the salt to the batch to be written
+                auto saltData = std::vector<uint8_t>(salt.cbegin(), salt.cend());
+                auto noCipher = Option<AESCipher>::NONE;
+                putPreferencesChange(batch, noCipher, PreferencesChange(PreferencesChangeType::PUT_TYPE, saltKey, saltData));
+
+                // iterate through all records already present and update them
+                auto it = std::unique_ptr<leveldb::Iterator>(_db->NewIterator(leveldb::ReadOptions()));
+                for (it->SeekToFirst(); it->Valid(); it->Next()) {
```

Comment:
Will this loop overwrite the salt data that was put before?

Reply:
I think it might. I’m refactoring a bit to put the salt stuff after the loop (I’m curious why the tests pass if the salt got overwritten). Nice catch, I’ll inquire.